File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-fx-5-4/FxOptionPremium.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class FxOptionPremium : NonNegativePayment
    {
        public FxOptionPremium(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode settlementInformationNode = xmlNode.SelectSingleNode("settlementInformation");
            
            if (settlementInformationNode != null)
            {
                if (settlementInformationNode.Attributes["href"] != null || settlementInformationNode.Attributes["id"] != null) 
                {
                    if (settlementInformationNode.Attributes["id"] != null) 
                    {
                        settlementInformationIDRef_ = settlementInformationNode.Attributes["id"].Value;
                        SettlementInformation ob = new SettlementInformation(settlementInformationNode);
                        IDManager.SetID(settlementInformationIDRef_, ob);
                    }
                    else if (settlementInformationNode.Attributes["href"] != null)
                    {
                        settlementInformationIDRef_ = settlementInformationNode.Attributes["href"].Value;
                    }
                    else
                    {
                        settlementInformation_ = new SettlementInformation(settlementInformationNode);
                    }
                }
                else
                {
                    settlementInformation_ = new SettlementInformation(settlementInformationNode);
                }
            }
            
        
            XmlNode quoteNode = xmlNode.SelectSingleNode("quote");
            
            if (quoteNode != null)
            {
                if (quoteNode.Attributes["href"] != null || quoteNode.Attributes["id"] != null) 
                {
                    if (quoteNode.Attributes["id"] != null) 
                    {
                        quoteIDRef_ = quoteNode.Attributes["id"].Value;
                        PremiumQuote ob = new PremiumQuote(quoteNode);
                        IDManager.SetID(quoteIDRef_, ob);
                    }
                    else if (quoteNode.Attributes["href"] != null)
                    {
                        quoteIDRef_ = quoteNode.Attributes["href"].Value;
                    }
                    else
                    {
                        quote_ = new PremiumQuote(quoteNode);
                    }
                }
                else
                {
                    quote_ = new PremiumQuote(quoteNode);
                }
            }
            
        
        }
        
    
        #region SettlementInformation_
        private SettlementInformation settlementInformation_;
        public SettlementInformation SettlementInformation_
        {
            get
            {
                if (this.settlementInformation_ != null)
                {
                    return this.settlementInformation_; 
                }
                else if (this.settlementInformationIDRef_ != null)
                {
                    settlementInformation_ = IDManager.getID(settlementInformationIDRef_) as SettlementInformation;
                    return this.settlementInformation_; 
                }
                else
                {
                      return this.settlementInformation_; 
                }
            }
            set
            {
                if (this.settlementInformation_ != value)
                {
                    this.settlementInformation_ = value;
                }
            }
        }
        #endregion
        
        public string settlementInformationIDRef_ { get; set; }
        #region Quote_
        private PremiumQuote quote_;
        public PremiumQuote Quote_
        {
            get
            {
                if (this.quote_ != null)
                {
                    return this.quote_; 
                }
                else if (this.quoteIDRef_ != null)
                {
                    quote_ = IDManager.getID(quoteIDRef_) as PremiumQuote;
                    return this.quote_; 
                }
                else
                {
                      return this.quote_; 
                }
            }
            set
            {
                if (this.quote_ != value)
                {
                    this.quote_ = value;
                }
            }
        }
        #endregion
        
        public string quoteIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-fx-5-4/FxOptionPremium.cpp
// FxOptionPremium.cpp 
#include "FxOptionPremium.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

FxOptionPremium::FxOptionPremium(TiXmlNode* xmlNode)
: NonNegativePayment(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //settlementInformationNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* settlementInformationNode = xmlNode->FirstChildElement("settlementInformation");

   if(settlementInformationNode){settlementInformationIsNull_ = false;}
   else{settlementInformationIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- settlementInformationNode , address : " << settlementInformationNode << std::endl;
   #endif
   if(settlementInformationNode)
   {
      if(settlementInformationNode->Attribute("href") || settlementInformationNode->Attribute("id"))
      {
          if(settlementInformationNode->Attribute("id"))
          {
             settlementInformationIDRef_ = settlementInformationNode->Attribute("id");
             settlementInformation_ = boost::shared_ptr<SettlementInformation>(new SettlementInformation(settlementInformationNode));
             settlementInformation_->setID(settlementInformationIDRef_);
             IDManager::instance().setID(settlementInformationIDRef_,settlementInformation_);
          }
          else if(settlementInformationNode->Attribute("href")) { settlementInformationIDRef_ = settlementInformationNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { settlementInformation_ = boost::shared_ptr<SettlementInformation>(new SettlementInformation(settlementInformationNode));}
   }

   //quoteNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* quoteNode = xmlNode->FirstChildElement("quote");

   if(quoteNode){quoteIsNull_ = false;}
   else{quoteIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- quoteNode , address : " << quoteNode << std::endl;
   #endif
   if(quoteNode)
   {
      if(quoteNode->Attribute("href") || quoteNode->Attribute("id"))
      {
          if(quoteNode->Attribute("id"))
          {
             quoteIDRef_ = quoteNode->Attribute("id");
             quote_ = boost::shared_ptr<PremiumQuote>(new PremiumQuote(quoteNode));
             quote_->setID(quoteIDRef_);
             IDManager::instance().setID(quoteIDRef_,quote_);
          }
          else if(quoteNode->Attribute("href")) { quoteIDRef_ = quoteNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { quote_ = boost::shared_ptr<PremiumQuote>(new PremiumQuote(quoteNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<SettlementInformation> FxOptionPremium::getSettlementInformation()
{
   if(!this->settlementInformationIsNull_){
        if(settlementInformationIDRef_ != ""){
             return boost::shared_static_cast<SettlementInformation>(IDManager::instance().getID(settlementInformationIDRef_));
        }else{
             return this->settlementInformation_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<SettlementInformation>();
   }
}
boost::shared_ptr<PremiumQuote> FxOptionPremium::getQuote()
{
   if(!this->quoteIsNull_){
        if(quoteIDRef_ != ""){
             return boost::shared_static_cast<PremiumQuote>(IDManager::instance().getID(quoteIDRef_));
        }else{
             return this->quote_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PremiumQuote>();
   }
}
}

